### PR TITLE
feat: Move `RoutedMessage` inside `Box` to reduce `PeerMessage` enum size

### PR DESCRIPTION
**Pull request by pmnoxx**
We should reduce size of `PeerMessage` enum to 64 bytes or less.
We can start by doing this for `PeerMessage::Routed` variant.

See #1313 

TODO
- [ ] variant `Challenge`: 1143 bytes                                                                                                                                                                                         
- [ ] variant `Transaction`: 287 bytes 
- [ ] variant `Handshake`: 215 bytes
- [ ] variant `HandshakeV2`: 207 bytes 
- [ ] variant `HandshakeFailure`: 127 bytes
- [ ] variant `RequestUpdateNonce`: 87 bytes

### CODE REVIEW

File: chain/network/src/network_protocol.rs [hunk 1]
```diff
@@ -424,3 +424,27 @@ impl PeerMessage {
...
+    }
+
+    #[test]
+    fn test_assert_size() {
+        assert_size!(HandshakeFailureReason, 64);
+        assert_size!(PeerMessage, 1144);
+        #[cfg(feature = "protocol_feature_routing_exchange_algorithm")]
+        assert_size!(RoutingSyncV2, 80);
+    }
+}
```
Comment (matklad) [inline chain/network/src/network_protocol.rs:450]: It might make sense to use `static_assert` crate here.

**Review by matklad — APPROVED** [review]

File: chain/network/src/network_protocol.rs (hunk 1 above)
Comment (pmnoxx) [reply chain/network/src/network_protocol.rs:450]: @matklad The best way would be to use `1.57.0` features 
```const _: () = assert!(std::mem::size_of::<u64>() == 8);```

Unfortunately, `const_format` is not fully stable yet. https://docs.rs/const_format/latest/const_format/
I can write a static assert like this:
```
assert!(std::mem::size_of::<u64>() == 8, "u64");
```
However, `assert!` doesn't support formatting in `static` case on rust stable.
I think it's best to add `TODO`, once that feature becomes stable.

Comment (matklad) [reply chain/network/src/network_protocol.rs:450]: Yeah, I mean that we can use this crate https://crates.io/crates/static_assertions which we already use, and which contains all the necessary polyfils. 

It was possible to "hand-write" this even before 1.57 (`[()][!cond]`), and I generally do perfer hand-writing such simple utilities, but, as we *already* depend on the crate, we might as use it. Though, no strong opinion here, the current approach is also fine!

Comment (pmnoxx) [reply chain/network/src/network_protocol.rs:450]: @matklad I looked into https://docs.rs/static_assertions/latest/static_assertions/
However, it's functionally the same as `1.57.0` static asserts. Neither of those examples contains proper prints on assert with format.

Comment (pmnoxx) [reply chain/network/src/network_protocol.rs]: Ok, I changed the code to use static asserts from `1.57.0`, they look good.

Comment (pmnoxx) [reply chain/network/src/network_protocol.rs]: https://doc.rust-lang.org/nightly/std/macro.const_format_args.html#

File: chain/network/src/network_protocol.rs [hunk 2]
```diff
@@ -424,3 +424,8 @@ impl PeerMessage {
         }
     }
 }
+
+const _: () =
+    assert!(std::mem::size_of::<HandshakeFailureReason>() <= 64, "HandshakeFailureReason");
+const _: () = assert!(std::mem::size_of::<PeerMessage>() <= 1144, "PeerMessage");
+const _: () = assert!(std::mem::size_of::<RoutingSyncV2>() <= 80, "RoutingSyncV2");
```
Comment (matklad) [inline chain/network/src/network_protocol.rs:431]: I'd probably put each of those after the corresponding struct. That way, when you look at the type's definiion, you realize that this is the place where we care about size.

**Comment by pmnoxx** [comment]
````
@mm-near @pmnoxx: /chain/network-primitives/ /chain/network/
```
@mm-near @pmnoxx Can you take a look?

**Comment by pmnoxx** [comment]
```
@mm-near: chain/network/ chain/network-primitives/
```
@mm-near Can you take a look?

**Comment by pmnoxx** [comment]
```
@mm-near: chain/network/ chain/network-primitives/
```
@mm-near Can you take a look?

File: chain/network/Cargo.toml [hunk 3]
```diff
@@ -4,7 +4,7 @@ version = "0.0.0"
 authors = ["Near Inc <hello@nearprotocol.com>"]
 publish = false
 # Please update rust-toolchain.toml as well when changing version here:
-rust-version = "1.56.0"
+rust-version = "1.57.0"
```
Comment (mm-near) [inline chain/network/Cargo.toml:7]: ? should this be a part of this PR? 

@matklad - can we update the rust version only for a single cargo? (as others still have 1.56, right?)

File: chain/network/src/network_protocol.rs [hunk 4]
```diff
@@ -267,6 +267,8 @@ pub enum HandshakeFailureReason {
     GenesisMismatch(GenesisId),
     InvalidTarget,
 }
+const _: () =
+    assert!(std::mem::size_of::<HandshakeFailureReason>() <= 64, "HandshakeFailureReason");
```
Comment (mm-near) [inline chain/network/src/network_protocol.rs:271]: add better comment for this assert (and others)

**Review by mm-near — CHANGES_REQUESTED** [review]

File: chain/network/Cargo.toml (hunk 3 above)
Comment (matklad) [reply chain/network/Cargo.toml:7]: We can, but it doesn't make sense. 

It also doesn't make sense that we have `1.57` in `rust-toolchain.toml` and `1.56` in all of the cargo-tomls. I guess we should have updated cargo.tomls as a part of https://github.com/near/nearcore/pull/5659. 

So:

* this change should be reverted from this PR
* we should send a separate PR to upgrade rust-version everywhere
  * it probably makes sense to fold this with https://github.com/near/nearcore/issues/6120
* we should patch https://github.com/near/nearcore/blob/6d86ccd4fee74c1567dc2d83d384b1952826e796/tools/themis/src/rules.rs#L28-L48 to also check that `rust-version` is the right one.

Comment (matklad) [reply chain/network/Cargo.toml:7]: filed https://github.com/near/nearcore/issues/6124

Comment (matklad) [reply chain/network/Cargo.toml:7]: To say this more explicitly: 

It's ok to actually use 1.57 features in this PR -- 1.57 is the version of Rust we use. It's a per-existing bug that Cargo.toml specifies older-than-actually-required rust version.

File: chain/network/src/peer/peer_actor.rs [hunk 5]
```diff
@@ -688,12 +688,16 @@ impl StreamHandler<Result<Vec<u8>, ReasonForBan>> for PeerActor {
...
-        {
-            self.txns_since_last_block.fetch_add(1, Ordering::AcqRel);
-        } else if let PeerMessage::Block(_) = &peer_msg {
+        if let PeerMessage::Routed(routed) = &peer_msg {
+            match routed.as_ref() {
+                RoutedMessage { body: RoutedMessageBody::ForwardTx(_), .. } => {
+                    self.txns_since_last_block.fetch_add(1, Ordering::AcqRel);
+                }
+                _ => {
+                    self.txns_since_last_block.store(0, Ordering::Release);
```
Comment (mm-near) [inline chain/network/src/peer/peer_actor.rs:697]: you changed the behavior here..


before the PR - we were 'storing' 0 only if we got a 'block' transaction. Now we're storing '0' on each non ForwardTX one.

**Review by mm-near — CHANGES_REQUESTED** [review]

File: chain/network/src/peer/peer_actor.rs (hunk 5 above)
Comment (pmnoxx) [reply chain/network/src/peer/peer_actor.rs:688]: @mm-near You are right, thanks. That was a good catch.

**Review by mm-near — APPROVED** [review]

**Pull request merged**